File: 250120/정수 n개의 합 3/sum-of-n-integers-3.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
    // Please write your code here.
    int arr[501][501];
    int prefix[501][501] = {};

    int n, k;
    cin >> n >> k;
    for (int i=1; i<=n; i++) {
        for (int j=1; j<=n; j++) {
            cin >> arr[i][j];
        }
    }

    for (int i=1; i<=n; i++) {
        for (int j=1; j<=n; j++) {
            prefix[i][j] = arr[i][j] + prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1];
        }
    }

    int max_val = 0;
    for (int i=1; i<=n-k+1; i++) {
        for (int j=1; j<=n-k+1; j++) {
            max_val = max(max_val, prefix[i+k-1][j+k-1] - prefix[i+k-1][j-1] - prefix[i-1][j+k-1] + prefix[i-1][j-1]);
        }
    }
    cout << max_val;

    return 0;
}

File: 250701/정수 N개의 합 3/sum-of-n-integers-3.py
N, K = map(int, input().split())

arr = [[0] * (N + 1)] + [[0] + list(map(int, input().split())) for _ in range(N)]

prefix = [[0 for j in range(N+1)] for i in range(N+1)]
for i in range(1, N+1):
    for j in range(1, N+1):
        prefix[i][j] = arr[i][j] + prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1]

M = 0
for i in range(K, N+1):
    for j in range(K, N+1):
        M = max(M, prefix[i][j] - prefix[i-K][j] - prefix[i][j-K] + prefix[i-K][j-K])

print(M)